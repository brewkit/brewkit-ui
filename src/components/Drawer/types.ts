import { ReactNode } from 'react';


export type Anchors = 'top' | 'right' | 'bottom' | 'left';


export interface Props {

    /**
     * Where to position the drawer
     */
    anchorFrom: Anchors,

    /**
     * Children passed to the component (JSX, String)
     */
    children?: ReactNode,

    /**
     * Custom class to provide component
     */
    className?: string,

    /**
     * Controlled prop to determine drawer open state
     */
    isOpen?: boolean,

    /**
     * Function passed to handle closing the drawer in a controlled state
     */
    onCloseDrawer: () => void,

}
