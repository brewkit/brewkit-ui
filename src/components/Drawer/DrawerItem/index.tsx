import React, { ReactElement } from 'react';
import clsx from 'clsx';
import DrawerContext from '../context';
import { Props } from './types';


/**
 * The DrawerItem component is used in a Drawer component that has the native ability to close the Drawer when clicked
 */
function DrawerItem({
    children,
    className,
    ...otherProps
}: Props): ReactElement {


    const { onCloseDrawer } = React.useContext(DrawerContext);


    const drawerItemClasses = clsx(
        'brew-Drawer__item',
        className,
    );


    return (
        <div className={drawerItemClasses} onClick={onCloseDrawer} {...otherProps}>
            {children}
        </div>
    );


}


export default DrawerItem;
