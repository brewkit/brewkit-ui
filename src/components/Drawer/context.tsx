import React from 'react';


interface ContextType {
    onCloseDrawer: () => void,
}


const DrawerContext = React.createContext<Partial<ContextType>>({});


export default DrawerContext;
