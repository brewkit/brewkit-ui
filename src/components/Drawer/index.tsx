import React, { ReactElement } from 'react';
import { createPortal } from 'react-dom';
import clsx from 'clsx';
import { Flipped, Flipper } from 'react-flip-toolkit';
import DrawerItem from './DrawerItem';
import DrawerContext from './context';
import { Props } from './types';


/**
 * The Drawer component is used when you want to have content slide out from either the Top, Right, Bottom, or Left
 */
function Drawer({
    anchorFrom,
    children,
    className,
    isOpen = false,
    onCloseDrawer,
    ...otherProps
}: Props): ReactElement {


    const drawerClasses = clsx(
        'brew-Drawer',
        `brew-Drawer--anchorFrom-${anchorFrom}`,
        { 'brew-Drawer--isOpen': isOpen },
        className,
    );


    return createPortal(
        (
            <DrawerContext.Provider value={{ onCloseDrawer }}>
                <Flipper flipKey={JSON.stringify([anchorFrom, className, isOpen, open])}>
                    <Flipped flipId="wrapper">
                        <div className={drawerClasses} {...otherProps}>
                            <span className="brew-Drawer__exit" onClick={onCloseDrawer}>&times;</span>
                            <div className="brew-Drawer__content">{children}</div>
                        </div>
                    </Flipped>
                    <div className="brew-Drawer__mask" onClick={onCloseDrawer} />
                </Flipper>
            </DrawerContext.Provider>
        ), document.body,
    );


}


Drawer.Item = DrawerItem;


export default Drawer;
