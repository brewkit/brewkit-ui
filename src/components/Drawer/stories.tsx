import React, { ReactElement } from 'react';
import { button, select } from '@storybook/addon-knobs';
import Drawer from './index';


export default {
    Component: Drawer,
    title: 'Components|Display/Drawer',
};


export const Default = (): ReactElement => {
    const anchorFrom = select('anchorFrom', ['top', 'right', 'bottom', 'left'], 'left');
    const [isOpen, setIsOpen] = React.useState(true);
    const handler = (): void => setIsOpen(!isOpen);
    button('Toggler', handler);

    return (
        <Drawer anchorFrom={anchorFrom} isOpen={isOpen} onCloseDrawer={handler}>
            <div>
                <div>Example Title</div>
                <div>Example SubTitle</div>
            </div>
            <div>item</div>
            <Drawer.Item>Example Drawer.Item</Drawer.Item>
        </Drawer>
    );
};
