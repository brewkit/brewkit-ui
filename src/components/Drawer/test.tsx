import React, { ReactNode } from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Drawer from './index';


jest.mock('react-dom', () => ({
    createPortal: (node: ReactNode): ReactNode => node,
}));


describe('Drawer', () => {


    it('Renders correctly', () => {
        const tree = renderer
            .create(<Drawer anchorFrom="left" onCloseDrawer={(): true => true}>test</Drawer>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Has default classes on all elements; Does not contain \'brew-Drawer--isOpen\'', () => {
        const drawer = mount(<Drawer anchorFrom="left" onCloseDrawer={(): true => true}>hello</Drawer>);
        const classes = [
            '.brew-Drawer',
            '.brew-Drawer--anchorFrom-left',
            '.brew-Drawer__content',
            '.brew-Drawer__exit',
            '.brew-Drawer__mask',
        ];

        classes.forEach((className: string) => expect(drawer.exists(className)).toBe(true));

        expect(drawer.exists('.brew-Drawer--isOpen')).toBe(false);
    });

    it('Attaches a custom class name to wrapper and modifies anchor class to right', () => {
        const drawer = shallow((
            <Drawer anchorFrom="right" className="brew-Drawer--custom" onCloseDrawer={(): true => true}>test</Drawer>
        ));

        expect(drawer.find('.brew-Drawer').hasClass('brew-Drawer--custom')).toBe(true);
        expect(drawer.find('.brew-Drawer').hasClass('brew-Drawer--anchorFrom-right')).toBe(true);
    });

    it('Has \'brew-Drawer--isOpen\' class, when isOpen prop is true', () => {
        const drawer = shallow(<Drawer anchorFrom="right" isOpen onCloseDrawer={(): true => true}>test</Drawer>);

        expect(drawer.find('.brew-Drawer').hasClass('brew-Drawer--isOpen')).toBe(true);
    });

    it('Contains \'brew-Drawer__item\' class when Drawer.Item component is used', () => {
        const drawer = mount((
            <Drawer anchorFrom="right" isOpen onCloseDrawer={(): true => true}>
                <Drawer.Item>Clickable</Drawer.Item>
            </Drawer>
        ));

        expect(drawer.find(Drawer.Item).childAt(0)
            .hasClass('brew-Drawer__item')).toBe(true);
    });


});
