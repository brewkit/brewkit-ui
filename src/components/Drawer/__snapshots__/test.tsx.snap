// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer Renders correctly 1`] = `
<div>
  <div
    className="brew-Drawer brew-Drawer--anchorFrom-left"
    data-flip-config="{\\"translate\\":true,\\"scale\\":true,\\"opacity\\":true}"
    data-flip-id="wrapper"
    data-portal-key="portal"
  >
    <span
      className="brew-Drawer__exit"
      onClick={[Function]}
    >
      Ã—
    </span>
    <div
      className="brew-Drawer__content"
    >
      test
    </div>
  </div>
  <div
    className="brew-Drawer__mask"
    onClick={[Function]}
  />
</div>
`;
